synapse/events/snapshot.py

```diff
@@ -254,6 +278,225 @@ async def get_prev_state_ids(
         )
 
 
+class UnpersistedEventContextForBatched(UnpersistedEventContextBase):
+    """
+    This is a version of an EventContext before the new state group (if any) has been
+    computed and stored. It contains information about the state before the event (which
+    also may be the information after the event, if the event is not a state event). The
+    UnpersistedEventContext must be converted into an EventContext by calling the method
+    'persist' on it before it is suitable to be sent to the DB for processing.
+
+        state_group_before_event:
+            the state group at/before the event. This is required if `for_batch` is True.
+
+        state_group:
+             this will always be None until it is persisted
+
+        state_group_before_event: The ID of the state group representing the state
+         of the room before this event.
+
+        state_delta_due_to_event: If the event is a state event, then this is the
+         delta of the state between `state_group` and `state_group_before_event`
+
+        prev_group: If it is known, ``state_group``'s prev_group.
+
+             If the event is a state event, this is normally the same as
+             ``state_group_before_event``.
+
+         delta_ids: If ``prev_group`` is not None, the state delta between ``prev_group``
+             and ``state_group``.
+
+         state_map: A map of the current state of the room
+    """
+
+    def __init__(
+        self,
+        storage_controller: "StorageControllers",
+        state_group: Optional[int],
+        state_delta_due_to_event: dict,
+        prev_group: Optional[int],
+        delta_ids: Optional[StateMap[str]],
+        partial_state: bool,
+        state_map: StateMap[str],
+        state_group_before_event: int,
+    ) -> None:
+        super().__init__(storage_controller)
+        self.state_group = state_group
+        self.state_delta_due_to_event = state_delta_due_to_event
+        self.prev_group = prev_group
+        self.delta_ids = delta_ids
+        self.partial_state = partial_state
+        self.state_map = state_map
+        self.state_group_before_event = state_group_before_event
```

Comment:
I'd be tempted to make this an `@attr.s` class and then we don't need the constructor?

Reply:
done, i think it's better?